### PR TITLE
Feature/docker support

### PR DESCRIPTION
**Pull request by schmunk42**
This is currently a draft, but a large part of the tests are already working.
I'd need more info about `redis` and `mongodb` tests and the local web-server to improve it further, also code-coverage can be enabled.

### CODE REVIEW

**Comment by DavertMik** [comment]
My only fear about this is that we will have tooo much great features in upcoming Codeception 2.2 :)

**Comment by Naktibalda** [comment]
why are you adding GitLab?

File: tests/web.suite.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ modules:
   enabled: [WebDriver, WebHelper]
   config:
     WebDriver:
+      host: firefox
```
Comment (Naktibalda) [inline tests/web.suite.yml]: This setting broke tests on Semaphore.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,40 @@
+FROM php:5.6-cli
```
Comment (sergeyklay) [inline Dockerfile]: we need use `TRAVIS_PHP_VERSION`, `PHP_VERSION`, etc

**Comment by Naktibalda** [comment]
This is cool and everything, but does it actually have to go into Codeception repo?
How about uploading it to Docker Hub instead and adding instruction to Docs?

There are 18 repositories having Codeception in their name in Docker Hub.
https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=Codeception&starCount=0

**Comment by schmunk42** [comment]
@Naktibalda 
> This is cool and everything, but does it actually have to go into Codeception repo?

I don't mind, @DavertMik WDYT?

> How about uploading it to Docker Hub instead and adding instruction to Docs?

For example, if you create an automated build from this repo, you can tag Docker images with corresponding versions ... and images are automatically build on DockerHub.
Tagging would have to be done manually when using a separate repo, like above.

> There are 18 repositories having Codeception in their name in Docker Hub.
https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=Codeception&starCount=0

Did you find a good one? I found them often outdated.
And I found none, which extend from the offical `php` image, which makes changes to the PHP version in testing very easy.

**Comment by schmunk42** [comment]
@Naktibalda
> why are you adding GitLab?

I use GitLab for testing, the runner which is used has `docker-compose` support - I didn't try this on travis yet. It could be removed, if travis works I think.

**Comment by DavertMik** [comment]
We had some discussion with @schmunk42 and here is the summary about this PR and the idea behind it.

* by adding Dockerfile we will get an official Codeception image built automatically on each release. Which sounds cool to me.
* we won't change current CI setup (travis.yml not changes, gitlab-ci not added)
* we will update parallel testing guides for using official docker image with settiing up selenium/chrome/firefox and other services like mysql/redis with docker-compose
* we can have our own `docker-compose.yml` for running our test suite locally. We should also update `tests/readme.md` to include that option

File: Dockerfile (hunk 2 above)
Comment (DavertMik) [inline Dockerfile]: this is not (should not be) related to travis setup.

File: tests/web.suite.yml (hunk 1 above)
Comment (DavertMik) [inline tests/web.suite.yml]: yes, please remove changes to semaphore setup

File: docs/12-ParallelExecution.md [hunk 3]
```diff
@@ -19,6 +19,47 @@ Docker works really well for isolating testing environments.
...
 
+### Docker
+
+> Section is under construction
+
+Requirements:
+
+ - `docker` or [Docker Toolbox](https://www.docker.com/products/docker-toolbox)
+
+Build the `codeception/codeception` image
```
Comment (DavertMik) [inline docs/12-ParallelExecution.md]: Better to have it already built

File: docs/12-ParallelExecution.md [hunk 4]
```diff
@@ -19,6 +19,47 @@ Docker works really well for isolating testing environments.
...
+
+Run single test
+    
+    docker-compose run --rm codecept run cli ExtensionsCest
+    
+Cleanup
+
+    docker-compose run --rm codecept clean
+
+In parallel
```
Comment (DavertMik) [inline docs/12-ParallelExecution.md]: what does `-p` option mean?

File: docs/12-ParallelExecution.md [hunk 5]
```diff
@@ -19,6 +19,47 @@ Docker works really well for isolating testing environments.
...
+
+Requirements:
+
+ - `docker` or [Docker Toolbox](https://www.docker.com/products/docker-toolbox)
+
+Build the `codeception/codeception` image
+
+    docker-compose build
+    
+Start
```
Comment (DavertMik) [inline docs/12-ParallelExecution.md]: Here you should mention about docker compose and how to enable all services for it

File: tests/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      - '80'
+  test1:
+    extends:
+      service: codecept
+
+  firefox:
+    image: selenium/standalone-firefox-debug:2.52.0
+  chrome:
+    image: selenium/standalone-chrome-debug:2.52.0
+  redis:
```
Comment (DavertMik) [inline tests/docker-compose.yml]: please also add mongo to it and instructions to tests/readme.md on how to run tests through docker compose

File: docs/12-ParallelExecution.md (hunk 5 above)
Comment (schmunk42) [inline docs/12-ParallelExecution.md]: I've added an Redis example.

File: docs/12-ParallelExecution.md (hunk 4 above)
Comment (schmunk42) [inline docs/12-ParallelExecution.md]: Using the long form now (project-name)

**Comment by schmunk42** [comment]
You can test the latest commit via `docker run schmunk42/codeception:develop`...

See also https://github.com/schmunk42/Codeception/blob/feature/docker-support/docs/12-ParallelExecution.md#using-codeception-docker-image

**Comment by schmunk42** [comment]
The docker support from this PR would be ready from my side.
Detailed docs are a separate thing, IMHO.

**Comment by DavertMik** [comment]
Thank you! I can probably handle the documentation part

**Comment by VeeeneX** [comment]
Hi guys, I'm just wondering if you can rename mounted volume from project to app, because [composer/composer](https://github.com/RobLoach/docker-composer) uses app folder, it's easier to use and it works with `volumes_from` option in docker-compose.

**Comment by schmunk42** [comment]
What are you trying to achieve? 
You are using two different images, you should not rely on the path in the container - why does it matter?

But you still might try to switch your `working_dir`.

**Comment by VeeeneX** [comment]
Application downloads its dependencies and then run codeception
Example of docker-compose:
```
  mailingoo-composer:
    image: composer/composer
    container_name: mailingoo-composer
    command: install --ignore-platform-reqs
    volumes_from:
      - mailingoo
    volumes:
      - ~/.ssh:/root/.ssh

  mailingoo-codeception:
    image: codeception/codeception
    container_name: mailingoo-codeception
    command: run
    volumes:
      - .:/project
    networks:
      - mailingoo
    depends_on:
      - mailingoo-composer
```

**Comment by schmunk42** [comment]
I see.

Could you try removing `volumes` from your codeception container and add `volumes_from` like you wanted it. And change `working_dir` to `/app`.

Can you provide a complete `docker-compose.yml` for testing this?

**Comment by VeeeneX** [comment]
This is example of docker-compose that I'm using for my application, I need to re-run some services after `up -d` because docker `depends_on` monitor only start of container not it's return status. But it's useful to have composer and all other services in one docker-compose.
```
version: '2'
services:
  mailingoo:
    container_name: mailingoo
    build:
      context: .
      dockerfile: docker/mailingoo/Dockerfile
    volumes:
      - .:/app
      - ./docker/mailingoo/php-ini-overrides.ini:/etc/php/7.0/fpm/conf.d/99-overrides.ini

  mailingoo-worker:
    container_name: mailingoo-worker
    build:
      context: .
      dockerfile: docker/mailingoo-worker/Dockerfile
    volumes_from:
      - mailingoo

  mailingoo-composer:
    image: composer/composer
    container_name: mailingoo-composer
    command: install --ignore-platform-reqs
    volumes_from:
      - mailingoo
    volumes:
      - ~/.ssh:/root/.ssh

  mailingoo-php-cs-fixer:
    image: ypereirareis/php-cs-fixer
    container_name: mailingoo-php-cs-fixer
    command: fix --level=psr2
    volumes_from:
      - mailingoo
    volumes:
      - ~/.ssh:/root/.ssh
    depends_on:
      - mailingoo-composer

  mailingoo-codeception:
    image: codeception/codeception
    container_name: mailingoo-codeception
    command: run
    volumes_from:
      - mailingoo
    depends_on:
      - mailingoo-composer
```

**Comment by schmunk42** [comment]
Can't run the example, because I've no Dockerfile from `mailingoo`, can I download it somewhere?

**Comment by VeeeneX** [comment]
@schmunk42 It's just an example, how application works. I just want to ask if you can rename /project mounted volume in dockerfile for Codeception to /app it's common usage for Docker

**Comment by schmunk42** [comment]
Should be possible.

**Pull request merged**